### ISSUE TITLE
pybind11 seems broken on MacOS/homebrew starting today?

### ISSUE THREAD

**Issue by lgritz** [pybind/pybind11#1929]
Just last night (basically, since Homebrew bumped to pybind11 v2.4.0), [my project](https://github.com/OpenImageIO/oiio) started failing with the following error message:

```
/usr/local/include/pybind11/numpy.h:157:23: error: expression is not an integral constant expression
        NPY_UINT64_ = platform_lookup<std::uint64_t, unsigned long, unsigned long long, unsigned int>(
                      ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/usr/local/include/pybind11/numpy.h:121:12: note: non-constexpr function 'get<0, int, int, int>' cannot be used in a constant expression
    return std::get<code_index::value>(std::make_tuple(codes...));
           ^

```

I verified that it builds properly with v2.3.0, v2.2.4 (my previous default that my build system downloads automatically if no system install is found). I also checked v2.4.1 and it exhibits the same failure.

Ring a bell? Is there something I should be fixing on my end?

**Comment by wjakob** [comment]
Probably this commit: https://github.com/pybind/pybind11/commit/4a3464fd88ae8666a670a17b99035beea6b464a8.

Could you provide some kind of reproducer? (Looks like it's potentially enough just to include numpy.h?) And some specifics about the machine please (compiler, OS).

cc @EricCousineau-TRI

**Comment by sizmailov** [comment]
Seems that the problem is that `std::get` is constexpr since from c++14: https://en.cppreference.com/w/cpp/utility/tuple/get

**Comment by wjakob** [comment]
Yes, I just came to the same conclusion (weird also, since pybind11's CI tests include C++11-only runs).

**Comment by lgritz** [comment]
Aha, somebody confused this with pybind14 :-)

**Comment by wjakob** [comment]
Working on a fix, just a moment.

**Comment by lgritz** [comment]
Thanks so much for the quick fix. In the mean time, my build system has the ability to auto-download pybind11 if a suitable version is not found on the system, so it's not hard for me to make a fallback so it doesn't break for users who have the affected versions and are still using C++11.

**Comment by wjakob** [comment]
Should be fixed on Master. Can you verify @lgritz? I plan to push out v2.4.2 after CI passes.

**Comment by wjakob** [comment]
Whoops, I forgot how limited constexpr is in C++11. (Can't have multiple return statements). Please try 7f5dad7d5fba3fb6fccef966f7cde5ca24509473.

**Comment by lgritz** [comment]
Confirmed, the current master works for me. Thanks!

**Comment by lgritz** [comment]
Your Travis matrix seem to have entries that use C++11, how that that not catch it in CI?

**Comment by wjakob** [comment]
I suspect that some compilers (or rather C++ library versions) are a bit to lenient. The fix is included in the just-released v2.4.2.

**Comment by lgritz** [comment]
Confirmed that v2.4.2 works for me on the same configurations where 2.4.0/2.4.1 failed. Thanks!
